third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp

```diff
@@ -181,14 +199,108 @@ struct InsertInstructionSchedHints
+    while (!convertOps.empty()) {
+      assert(convertOps.size() == 1);
+      value = convertOps.begin()->getResult();
+      convertOps = getUsersOfType<triton::gpu::ConvertLayoutOp>(value);
+    }
+    return value;
+  }
+
+  void anotateDotOperands(scf::ForOp forOp) const {
+    llvm::SmallVector<triton::LoadOp> loadOps;
+    forOp.walk(
+        [&loadOps](triton::LoadOp loadOp) { loadOps.push_back(loadOp); });
+
+    ValueRange yieldedValues = forOp.getYieldedValues();
+    auto initArgs = forOp.getRegionIterArgs();
+
+    MLIRContext *ctx = forOp->getContext();
+    mlir::OpBuilder rewriter(ctx);
+
+    for (auto loadOp : loadOps) {
+      Value loadResult = loadOp.getResult();
+
+      // rewind till the first carried loop iteration regarding `tt.load`
+      Value loopCarriedLoadValue = loadResult;
+      bool foundFirstCarriedLoopIteration = false;
+      while (!foundFirstCarriedLoopIteration) {
+        auto it = llvm::find(yieldedValues, loopCarriedLoadValue);
+        if (it != yieldedValues.end()) {
+          size_t idx = std::distance(yieldedValues.begin(), it);
+          loopCarriedLoadValue = initArgs[idx];
+        } else {
+          foundFirstCarriedLoopIteration = true;
+        }
+      }
+
+      loopCarriedLoadValue = rewindConvertLayoutOps(loopCarriedLoadValue);
+      assert(loopCarriedLoadValue.hasOneUse());
+
+      // handle pipelining - i.e., `local_store`, `memdesc_subview`,
+      // `local_load` ops
+      triton::gpu::LocalLoadOp localLoadOp = nullptr;
+      auto loadOpUser = *(loopCarriedLoadValue.getUsers().begin());
+      auto localStoreOp = llvm::dyn_cast<triton::gpu::LocalStoreOp>(loadOpUser);
+      if (localStoreOp) {
+        auto subviewOp = llvm::cast<triton::gpu::MemDescSubviewOp>(
+            localStoreOp->getOperand(1).getDefiningOp());
+        Value subviewResult = subviewOp.getResult();
+        auto it = llvm::find(yieldedValues, subviewResult);
+        assert(it != yieldedValues.end());
+        size_t idx = std::distance(yieldedValues.begin(), it);
+        Value loopCarriedSubviewValue = initArgs[idx];
+
+        auto subviewLoadOps =
+            getUsersOfType<triton::gpu::LocalLoadOp>(loopCarriedSubviewValue);
+        assert(subviewLoadOps.size() == 1);
+        localLoadOp = *(subviewLoadOps.begin());
+
+        loopCarriedLoadValue = localLoadOp.getResult();
+        loopCarriedLoadValue = rewindConvertLayoutOps(loopCarriedLoadValue);
+      }
+
+      // find the corresponding `DotOp`
+      auto dots = getUsersOfType<triton::DotOp>(loopCarriedLoadValue);
+      assert(dots.size() == 1);
+
+      // find which `DotOp` argument the current `loadOp` belongs to
+      auto dotOperands = dots.begin()->getOperands();
+      auto it = llvm::find(dotOperands, loopCarriedLoadValue);
+      assert(it != dotOperands.end());
+      size_t opIdx = std::distance(dotOperands.begin(), it);
+
+      // set `idxOp` attributes
+      auto nameAttr = str_attr("idxOp");
```

Comment:
Avoid hard-coded strings - i.e., `"idxOp"`. Add a proper MLIR attribute. Check attributes of `triton::gpu.local_load`. Maybe there is already one which can be re-used.

Reply:
This is fine actually--I don't see we have a story to add a builtin attribute to the op itself for semantics ; this is meant to carry information between different passes.

Reply:
Agree. But I don't like hard-coded strings because it looks like some sort of magical variables, especially when it is spread over several passes

Reply:
Fair enough. It's not a big issue; so your call. :)